cmd/registry-replacer/main.go

```diff
@@ -606,6 +627,117 @@ func pruneReplacements(config *api.ReleaseBuildConfiguration, filter asDirective
 	return utilerrors.NewAggregate(errs)
 }
 
+// pruneUnusedBaseImages uses the fully-resolved config to make sure an image is not used directly in  the config, or within any of the tests.
+// If it is not, then we prune it.
+func pruneUnusedBaseImages(config *api.ReleaseBuildConfiguration, resolvedConfig *api.ReleaseBuildConfiguration) error {
+	usedBaseImages := sets.NewString()
+
+	getOperatorImages(config, usedBaseImages)
+	for _, step := range resolvedConfig.RawSteps {
+		switch {
+		case step.BundleSourceStepConfiguration != nil:
+			getBundleSourceImages(&usedBaseImages, step.BundleSourceStepConfiguration)
+		case step.OutputImageTagStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.OutputImageTagStepConfiguration.From))
+		case step.PipelineImageCacheStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.PipelineImageCacheStepConfiguration.From))
+		case step.RPMImageInjectionStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.RPMImageInjectionStepConfiguration.From))
+		case step.SourceStepConfiguration != nil:
+			usedBaseImages.Insert(string(step.SourceStepConfiguration.From))
+		case step.TestStepConfiguration != nil:
+			getTestStepImages(resolvedConfig, &usedBaseImages, step.TestStepConfiguration)
+		case step.IndexGeneratorStepConfiguration != nil || step.ProjectDirectoryImageBuildStepConfiguration != nil || step.ProjectDirectoryImageBuildInputs != nil ||
+			step.ReleaseImagesTagStepConfiguration != nil || step.ResolvedReleaseImagesStepConfiguration != nil || step.RPMServeStepConfiguration != nil:
+			// no op
+		default:
+			return fmt.Errorf("unsupported step configuration provided when pruning base images")
+		}
+	}
+
+	for _, test := range resolvedConfig.Tests {
+		getTestStepImages(resolvedConfig, &usedBaseImages, &test)
+	}
+
+	for _, image := range resolvedConfig.Images {
+		usedBaseImages.Insert(string(image.From))
+		for input := range image.Inputs {
+			usedBaseImages.Insert(input)
+		}
+	}
+
+	for sourceImage := range config.BaseImages {
+		var keep bool
+		for _, candidate := range usedBaseImages.List() {
```

Comment:
This is the same as iterating over the keys, it doesn't need to construct a temporary slice.